### ISSUE TITLE
Write more tests

### ISSUE THREAD

**Issue by lbischof** [antoiner77/caddy-ansible#62]
We should write more tests. These should be compatible with Travis and Vagrant. Maybe we could change `test.yml` to use `include_role` multiple times with different variables set. We could also add tasks that check the state at the end. How would we check idempotency? Any ideas are welcome

edit: This looks interesting https://github.com/metacloud/molecule

**Comment by paulfantom** [comment]
If you want I can quickly create whole CI with molecule (unfortunately 1.x, not 2.x), travisCI and docker if you want. I have done something similar here: https://github.com/cloudalchemy/ansible-prometheus. Basically it run tests on 3 ansible versions, 2 app versions and 5 OSes.

**Comment by lbischof** [comment]
That would be awesome. If I understand it correctly it would be possible
for multiple tests with different variables?

**Comment by paulfantom** [comment]
In general it comes down to writing an ansible playbook (like [this one](https://github.com/cloudalchemy/ansible-node-exporter/blob/master/tests/playbook.yml)) and then checking it's outcome with [testinfra](testinfra.readthedocs.io).

Longer explanation:
I'm using matrix of parallel builds (2x3) in travis since travis can pass environment variables into build machine. So something like this: 
```yaml
env:
  - ANSIBLE=2.2.3 PROMETHEUS=1.8.2
  - ANSIBLE=2.2.3 PROMETHEUS=2.0.0
  - ANSIBLE=2.3.2 PROMETHEUS=1.8.2
  - ANSIBLE=2.3.2 PROMETHEUS=2.0.0
  - ANSIBLE=2.4.2 PROMETHEUS=1.8.2
  - ANSIBLE=2.4.2 PROMETHEUS=2.0.0
```
gives you 6 parallel pipelines.
Next stage happens inside build VM and is managed by molecule itself. Molecule creates couple of docker containers and runs ansible playbook against them (it treats containers as different hosts from ansible inventory). It provisions instances, runs ansible playbook twice (idempotence tests!), lints code and checks infrastructure with testinfra.
Basically this gives an opportunity to test whatever you want :smile: 

Also if you want I can include a [simple script](https://github.com/cloudalchemy/ansible-prometheus/blob/master/.travis/generatetag.sh) to auto-tag every merge to master branch according to semver.

**Comment by lbischof** [comment]
Sounds great. Would be really cool, if you have time. I'd prefer to tag
manually though, since there aren't that many changes. Maybe in the future.

**Comment by paulfantom** [comment]
Ok, I just did it :smile: 
However CI pipeline will fail due to:

```yaml
--> Executing ansible-lint...
[ANSIBLE0006] systemctl used in place of systemd module
/home/paulfantom/Development/ansible_roles/caddy-ansible/handlers/main.yml:6
Task/Handler: Reload systemd service file

[ANSIBLE0016] Tasks that run when changed should likely be handlers
/home/paulfantom/Development/ansible_roles/caddy-ansible/tasks/main.yml:32
Task/Handler: Download Caddy

[ANSIBLE0016] Tasks that run when changed should likely be handlers
/home/paulfantom/Development/ansible_roles/caddy-ansible/tasks/main.yml:49
Task/Handler: Download Caddy PGP key

[ANSIBLE0016] Tasks that run when changed should likely be handlers
/home/paulfantom/Development/ansible_roles/caddy-ansible/tasks/main.yml:55
Task/Handler: Download Caddy signature

[ANSIBLE0016] Tasks that run when changed should likely be handlers
/home/paulfantom/Development/ansible_roles/caddy-ansible/tasks/main.yml:59
Task/Handler: Verify Caddy signature

[ANSIBLE0016] Tasks that run when changed should likely be handlers
/home/paulfantom/Development/ansible_roles/caddy-ansible/tasks/main.yml:63
Task/Handler: Extract Caddy

[ANSIBLE0010] Package installs should not use latest
tests/playbook.yml:5
Task/Handler: Install curl
```

**Comment by paulfantom** [comment]
If you think everything here is ok, I can add `skip_ansible_lint` tag to those tasks and it will skip linting them.

**Comment by lbischof** [comment]
I merged #68 and will leave this issue open. `ANSIBLE0006` could be fixed with the `systemd` module, but  we also currently support Upstart. It seems forcing a daemon-reload has only been added in 2.4 (see examples on module page), so conditionally using `service` or `systemd` is probably the better choice.

**Comment by LorenzBischof** [comment]
Closing because the repository has moved to https://github.com/caddy-ansible/caddy-ansible If this is still an issue, please reopen the issue in the new repository.
